src/query/storages/share/src/share.rs

```diff
@@ -0,0 +1,238 @@
+pub struct ShareDatabaseSpec {
+    pub name: String,
+    pub id: u64,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct ShareSpec {
+    pub name: String,
+    pub database: Option<ShareDatabaseSpec>,
+    pub tables: Vec<ShareTableSpec>,
+    pub tenants: Vec<String>,
+}
+
+impl ShareSpec {
+    pub fn new(name: String) -> Self {
+        ShareSpec {
+            name,
+            database: None,
+            tables: vec![],
+            tenants: vec![],
+        }
+    }
+
+    pub fn grant_object(
+        &mut self,
+        name: &ShareGrantObjectName,
+        id: &u64,
+        options: &Option<BTreeMap<String, String>>,
+    ) -> Self {
+        match name {
+            ShareGrantObjectName::Database(name) => {
+                self.database = Some(ShareDatabaseSpec {
+                    name: name.clone(),
+                    id: *id,
+                });
+                self.clone()
+            }
+            ShareGrantObjectName::Table(_db, table_name) => {
+                if let Some(options) = options {
+                    let location = options
+                        .get(OPT_KEY_SNAPSHOT_LOCATION)
+                        // for backward compatibility, we check the legacy table option
+                        .or_else(|| options.get(OPT_KEY_LEGACY_SNAPSHOT_LOC))
+                        .cloned();
+                    if let Some(location) = location {
+                        self.tables
+                            .push(ShareTableSpec::new(table_name, *id, location));
+                        self.tables.sort_by(|a, b| a.name.cmp(&b.name));
+                    }
+                }
+
+                self.clone()
+            }
+        }
+    }
+
+    pub fn revoke_object(&mut self, name: &ShareGrantObjectName) -> Self {
+        match name {
+            ShareGrantObjectName::Database(_name) => {
+                self.database = None;
+                self.clone()
+            }
+            ShareGrantObjectName::Table(_db, table_name) => {
+                let index = self.tables.binary_search_by(|a| a.name.cmp(table_name));
+                if let Ok(index) = index {
+                    self.tables.remove(index);
+                }
+                self.clone()
+            }
+        }
+    }
+}
+
+pub enum ModShareSpec {
+    CreateShare {
+        share_name: String,
+    },
+    DropShare,
+    GrantShareObject {
+        object_name: ShareGrantObjectName,
+        id: u64,
+        options: Option<BTreeMap<String, String>>,
+    },
+    RevokeShareObject {
+        object_name: ShareGrantObjectName,
+    },
+    AddShareAccounts {
+        accounts: Vec<String>,
+    },
+    RemoveShareAccounts {
+        accounts: Vec<String>,
+    },
+}
+
+impl ModShareSpec {
+    pub async fn save(&self, tenant: String, share_id: u64, operator: Operator) -> Result<()> {
+        let location = format!("{}/{}/{}_share.json", SHARE_CONFIG_PREFIX, tenant, share_id);
```

Comment:
we can save all the shares' info into one file? like `tenant/shares.json`.

the config would not become too big, as we can introduce some quotas on limit of shares on the higher level.